### PR TITLE
Use syntax updated with Python 3.7 and avoid unnecessary list creations

### PR DESCRIPTION
**Pull request by bryant1410**
I ran the following, which updates the code to syntax to Python >= 3.7:

```bash
pyupgrade $(find . -name "*.py" -type f) --py37-plus
```

This includes, for example, the Python-2 style of calling `super` and using f-strings where possible. I reviewed these changes manually.

Also, I looked for unnecessary list comprehension creations, where a generator expression could have been created instead, thus saving memory and compute (though small if the lists were small).

### CODE REVIEW

File: hydra/core/config_store.py [hunk 1]
```diff
@@ -133,10 +133,10 @@ def get_type(self, path: str) -> ObjectType:
     def list(self, path: str) -> List[str]:
         d = self._open(path)
         if d is None:
-            raise IOError(f"Path not found {path}")
+            raise OSError(f"Path not found {path}")
 
         if not isinstance(d, dict):
-            raise IOError(f"Path points to a file : {path}")
+            raise OSError(f"Path points to a file : {path}")
```
Comment (Jasha10) [inline hydra/core/config_store.py:139]: Why is this change necessary?

File: hydra/test_utils/test_utils.py [hunk 2]
```diff
@@ -221,7 +221,7 @@ def find_parent_dir_containing(
         cur = os.path.relpath(os.path.join(cur, ".."))
         max_up = max_up - 1
     if max_up == 0:
-        raise IOError(f"Could not find {target} in parents of {os.getcwd()}")
+        raise OSError(f"Could not find {target} in parents of {os.getcwd()}")
```
Comment (Jasha10) [inline hydra/test_utils/test_utils.py:224]: Same question.

File: tools/configen/configen/utils.py [hunk 3]
```diff
@@ -59,10 +59,7 @@ def type_str(t: Any) -> str:
 
 def is_tuple_annotation(type_: Any) -> bool:
     origin = getattr(type_, "__origin__", None)
-    if sys.version_info < (3, 7, 0):
-        return origin is Tuple or type_ is Tuple  # pragma: no cover
-    else:
-        return origin is tuple  # pragma: no cover
+    return origin is tuple  # pragma: no cover
```
Comment (Jasha10) [inline tools/configen/configen/utils.py:62]: I think `# pragma: no cover` comment here can be removed.
The comment was only there to prevent our coverage linter from complaining when the line was not hit by the python3.6 test suite.

File: tools/configen/tests/test_modules/future_annotations.py [hunk 4]
```diff
@@ -26,12 +26,12 @@ class ExampleClass:
...
-        passthrough_list: List[LibraryClass],
-        dataclass_val: List[User],
-        def_value: List[str] = [],
+        lst: list[str],
+        passthrough_list: list[LibraryClass],
+        dataclass_val: list[User],
+        def_value: list[str] = [],
         default_str="Bond, James Bond",
-        none_str: Optional[str] = None,
+        none_str: str | None = None,
```
Comment (Jasha10) [inline tools/configen/tests/test_modules/future_annotations.py:34]: This change is not compatible with python3.7.

Python3.7 doesn't support subscripting the builtin `list`, nor does it support using the bar `|` for union types.

**Review by Jasha10 — CHANGES_REQUESTED** [review]
Thanks @bryant1410.
Overall it looks good, with a few issues:

File: tools/configen/tests/test_modules/future_annotations.py (hunk 4 above)
Comment (bryant1410) [reply tools/configen/tests/test_modules/future_annotations.py:34]: It does for typing (but e.g. not type aliases) when using `from __future__ import annotations` as at the beggining of this file.

Comment (bryant1410) [reply tools/configen/tests/test_modules/future_annotations.py:34]: This is because the usage of that annotation makes the types be treated as strings and they are evaluated later.

File: hydra/core/config_store.py (hunk 1 above)
Comment (bryant1410) [reply hydra/core/config_store.py:139]: It's not strictly necessary because they are semantically the same, but it's the recommended canonical name now. See https://stackoverflow.com/a/29347946/1165181

**Comment by bryant1410** [comment]
BTW, the tests are failing for another reason, right?

**Comment by Jasha10** [comment]
> BTW, the tests are failing for another reason, right?

I think so. The failing test should be fixed by PR #2677, which was recently merged into `main`. Could you please rebase this PR against the `main` branch?

File: tools/configen/tests/test_modules/future_annotations.py (hunk 4 above)
Comment (Jasha10) [reply tools/configen/tests/test_modules/future_annotations.py:34]: I wonder, can we get away with removing the `from __future__ import annotations` line in this file?

File: hydra/core/config_store.py (hunk 1 above)
Comment (Jasha10) [reply hydra/core/config_store.py:139]:
```python
>>> IOError is OSError
True
```
I see. Thanks.

**Comment by bryant1410** [comment]
> I think so. The failing test should be fixed by PR #2677, which was recently merged into `main`. Could you please rebase this PR against the `main` branch?

Okay! I just merged instead of rebasing, hope it's fine. You can always rebase everything at the end, and I think it should remove the merge-type commits.

File: tools/configen/tests/test_modules/future_annotations.py (hunk 4 above)
Comment (bryant1410) [reply tools/configen/tests/test_modules/future_annotations.py:34]: It's called `future_annotations`, so sounds like no? 😄

**Comment by bryant1410** [comment]
BTW, I see the latest in `main` is still failing. You sure the tests are gonna work?

File: tests/instantiate/__init__.py [hunk 5]
```diff
@@ -39,7 +39,7 @@ def partial_equal(obj1: Any, obj2: Any) -> bool:
     if isinstance(obj1, list):
         if len(obj1) != len(obj2):
             return False
-        return all([partial_equal(obj1[i], obj2[i]) for i in range(len(obj1))])
+        return all(partial_equal(o1, o2) for o1, o2 in zip(obj1))
```
Comment (Jasha10) [inline tests/instantiate/__init__.py:42]: There's an error in this update to the list comprehension.
Should be something like this:
```python
        return all(partial_equal(o1, o2) for o1, o2 in zip(obj1, obj2))
```

**Review by Jasha10 — CHANGES_REQUESTED** [review]
Hmm, no. It seems there's an issue with the diff. See comment below:

File: tools/configen/tests/test_modules/future_annotations.py (hunk 4 above)
Comment (Jasha10) [reply tools/configen/tests/test_modules/future_annotations.py:34]: Oh, great point.

File: tests/instantiate/__init__.py (hunk 5 above)
Comment (bryant1410) [reply tests/instantiate/__init__.py:42]: Oh, thank you! For context, I did that one manually, not with the tool.

File: tools/configen/tests/test_modules/future_annotations.py (hunk 4 above)
Comment (Jasha10) [reply tools/configen/tests/test_modules/future_annotations.py:34]: To keep consistency with the rest of the codebase, I still think there's benefit to using `List` instead of `list`. We don't need to convert `List -> list` just because the pyupgrade tool said so.

**Comment by Jasha10** [comment]
Currently three CI jobs are failing on this branch:
* lint-3.10: seems unrelated to this diff
* test_tools-3.10: related
* test_core-3.10: seems unrelated to this diff (see #2692)

I believe the test_tools failure is caused by this diff's change to `tools/configen/tests/test_modules/future_annotations.py`.

**Review by Jasha10 — APPROVED** [review]
Looks good. Thanks @bryant1410!

**Review by shagunsodhani — APPROVED** [review]
